File: app/src/main/java/com/vistony/salesforce/Dao/Retrofit/HeaderDispatchSheetRepository.java
package com.vistony.salesforce.Dao.Retrofit;

import static com.vistony.salesforce.Controller.Utilitario.Utilitario.getDateTime;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.vistony.salesforce.Controller.Retrofit.Api;
import com.vistony.salesforce.Controller.Retrofit.Config;
import com.vistony.salesforce.Dao.SQLite.DetailDispatchSheetSQLite;
import com.vistony.salesforce.Dao.SQLite.HeaderDispatchSheetSQLite;
import com.vistony.salesforce.Dao.SQLite.ParametrosSQLite;
import com.vistony.salesforce.Entity.Retrofit.Modelo.InvoicesEntity;
import com.vistony.salesforce.Entity.Retrofit.Respuesta.HeaderDispatchSheetEntityResponse;
import com.vistony.salesforce.Entity.SesionEntity;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class HeaderDispatchSheetRepository extends ViewModel {


    private MutableLiveData<String> status= new MutableLiveData<>();

    public MutableLiveData<String> getAndInsertHeaderDispatchSheet(
            String Imei
            , String FechaDespacho
            ,Context context
    ){
        status= new MutableLiveData<>();
        //if(SesionEntity.perfil_id.equals("Chofer")||SesionEntity.perfil_id.equals("CHOFER"))
        //{
            Config.getClient().create(Api.class)
                    .getHeaderDispatchSheet(
                            Imei
                            , FechaDespacho
                    ).enqueue(new Callback<HeaderDispatchSheetEntityResponse>() {
                        @Override
                        public void onResponse(Call<HeaderDispatchSheetEntityResponse> call, Response<HeaderDispatchSheetEntityResponse> response) {
                            Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-call" + call.toString());
                            HeaderDispatchSheetEntityResponse headerDispatchSheetEntityResponse = response.body();
                            Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-response" + response.toString());
                            if (response.isSuccessful() && headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().size() > 0) {
                                Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity()" + headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().toString());

                                HeaderDispatchSheetSQLite headerDispatchSheetSQLite = new HeaderDispatchSheetSQLite(context);
                                ParametrosSQLite parametrosSQLite = new ParametrosSQLite(context);
                                headerDispatchSheetSQLite.ClearTableHeaderDispatchDate();
                                DetailDispatchSheetSQLite detailDispatchSheetSQLite = new DetailDispatchSheetSQLite(context);
                                detailDispatchSheetSQLite.ClearTableDetailDispatchSheet();
                                headerDispatchSheetSQLite.InsertHeaderDispatchSheet(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity(), FechaDespacho);
                                for (int i = 0; i < headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().size(); i++) {
                                    //Documentos
                                    if (headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch() == null || headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().size() == 0) {
                                        //ObjCliente.setListInvoice(null);
                                    } else {

                                        for (int j = 0; j < headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().size(); j++) {
                                            headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().get(j).setControlid(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getControl_id());
                                        }
                                        detailDispatchSheetSQLite.InsertDetailDispatchSheet(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch(), FechaDespacho);

                                    }
                                }
                                Integer countHeadDispatchSheet = headerDispatchSheetSQLite.getCountHeaderDispatchSheet();
                                parametrosSQLite.ActualizaCantidadRegistros("19", "HOJA DESPACHO", "" + countHeadDispatchSheet, getDateTime());
                                Integer countDetailDispatchSheet = detailDispatchSheetSQLite.getCountDetailDispatchSheet();
                                parametrosSQLite.ActualizaCantidadRegistros("20", "HOJA DESPACHO DETALLE", "" + countDetailDispatchSheet, getDateTime());
                                status.setValue("1");
                            } else {
                                status.setValue("0");
                            }


                        }

                        @Override
                        public void onFailure(Call<HeaderDispatchSheetEntityResponse> call, Throwable t) {
                            status.setValue("0");
                            Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-error" + t.toString());
                        }
                    });
        return status;
    }

    public MutableLiveData<String> getAndInsertHeaderDispatchSheetSalesPerson(
            String Imei
            , String FechaDespacho
            ,Context context
    ){
        status= new MutableLiveData<>();
        Config.getClient().create(Api.class)
                .getHeaderDispatchSheetSalesPerson(
                        Imei
                        , FechaDespacho
                        ,"Seller"
                ).enqueue(new Callback<HeaderDispatchSheetEntityResponse>() {
                    @Override
                    public void onResponse(Call<HeaderDispatchSheetEntityResponse> call, Response<HeaderDispatchSheetEntityResponse> response) {
                        Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-call" + call.toString());
                        HeaderDispatchSheetEntityResponse headerDispatchSheetEntityResponse = response.body();
                        Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-response" + response.toString());
                        if (response.isSuccessful() && headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().size() > 0) {
                            Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity()" + headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().toString());

                            HeaderDispatchSheetSQLite headerDispatchSheetSQLite = new HeaderDispatchSheetSQLite(context);
                            ParametrosSQLite parametrosSQLite = new ParametrosSQLite(context);
                            headerDispatchSheetSQLite.ClearTableHeaderDispatchDate();
                            DetailDispatchSheetSQLite detailDispatchSheetSQLite = new DetailDispatchSheetSQLite(context);
                            detailDispatchSheetSQLite.ClearTableDetailDispatchSheet();
                            headerDispatchSheetSQLite.InsertHeaderDispatchSheet(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity(), FechaDespacho);
                            for (int i = 0; i < headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().size(); i++) {
                                //Documentos
                                if (headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch() == null || headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().size() == 0) {
                                    //ObjCliente.setListInvoice(null);
                                } else {

                                    for (int j = 0; j < headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().size(); j++) {
                                        headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch().get(j).setControlid(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getControl_id());
                                    }
                                    detailDispatchSheetSQLite.InsertDetailDispatchSheet(headerDispatchSheetEntityResponse.getHeaderDispatchSheetEntity().get(i).getListDedailDispatch(), FechaDespacho);

                                }
                            }
                            Integer countHeadDispatchSheet = headerDispatchSheetSQLite.getCountHeaderDispatchSheet();
                            parametrosSQLite.ActualizaCantidadRegistros("19", "HOJA DESPACHO", "" + countHeadDispatchSheet, getDateTime());
                            Integer countDetailDispatchSheet = detailDispatchSheetSQLite.getCountDetailDispatchSheet();
                            parametrosSQLite.ActualizaCantidadRegistros("20", "HOJA DESPACHO DETALLE", "" + countDetailDispatchSheet, getDateTime());
                            status.setValue("1");
                        } else {
                            status.setValue("0");
                        }


                    }

                    @Override
                    public void onFailure(Call<HeaderDispatchSheetEntityResponse> call, Throwable t) {
                        status.setValue("0");
                        Log.e("REOS", "HeaderDispatchSheetRepository-getAndInsertHeaderDispatchSheet-error" + t.toString());
                    }
                });
        return status;
    }
}

File: app/src/main/java/com/vistony/salesforce/kotlin/data/HeaderDispatchSheetRepository.kt
package com.vistony.salesforce.kotlin.data

import android.content.Context
import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.vistony.salesforce.Entity.SesionEntity
import com.vistony.salesforce.kotlin.api.RetrofitApi
import com.vistony.salesforce.kotlin.api.RetrofitConfig
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/*
interface HeaderDispatchSheetRepositorykt {

    suspend fun getStatusDispatchSheet(Imei: String,FechaDespacho: String,context: Context): MutableLiveData<String>
}*/


class HeaderDispatchSheetRepository(
    )
{
    private val _status = MutableLiveData<String>()
    val status: MutableLiveData<String> = _status

    suspend fun getStateDispatchSheet(Imei:String,FechaDespacho:String,context:Context)
    //: MutableLiveData<String>
    {
        //_status.value =dispatchRepository.getDispatch(Imei, FechaDespacho)!!
        // _dispatch.value = result
        Log.e(
            "REOS",
            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-Imei"+Imei
        )
        Log.e(
            "REOS",
            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-FechaDespacho"+FechaDespacho
        )
        try {
            val retrofitConfig: RetrofitConfig? = RetrofitConfig()
            val service = retrofitConfig?.getClientLog()?.create(
                RetrofitApi
                ::class.java
            )

            Log.e(
                "REOS",
                "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-service" + service.toString()
            )
            service?.getHeaderDispatchSheet(
                Imei, FechaDespacho
            )?.enqueue(object : Callback<DispatchSheetResponse?> {

                override fun onResponse(
                    call: Call<DispatchSheetResponse?>,
                    response: Response<DispatchSheetResponse?>
                ) {
                    Log.e(
                        "REOS",
                        "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-call$call"
                    )

                    Log.e(
                        "REOS",
                        "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response$response"
                    )

                        Log.e(
                            "REOS",
                            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response.isSuccessful"
                        )
                        val headerDispatchSheetResponse = response.body()
                        //headerDispatchSheetResponse?.getDispatchSheetEntity()?.size
                        Log.e(
                            "REOS",
                            "HeaderDispatchSheetRepositorykt?.getStateDispatchSheet()?.size" + headerDispatchSheetResponse?.getDispatchSheetEntity()?.size
                        )

                    if (response.isSuccessful&&headerDispatchSheetResponse?.getDispatchSheetEntity()?.size!!>0) {
                        val sdf = SimpleDateFormat("yyyy-MM-dd hh:mm:ss")
                        val calendario = GregorianCalendar()
                        for (i in 0..headerDispatchSheetResponse?.getDispatchSheetEntity()?.size!! - 1) {
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).control_id
                            Log.e(
                                "REOS",
                                "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-headerDispatchSheetResponse?.getDispatchSheetEntity()?.get(i).docEntry" + headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                    .get(i).control_id
                            )
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).compania_id=SesionEntity.compania_id
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).fuerzatrabajo_id=SesionEntity.fuerzatrabajo_id
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).usuario_id=SesionEntity.usuario_id
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).fechahojadespacho=FechaDespacho
                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                .get(i).Datetimeregister=sdf.format(calendario.getTime())
                            for (j in 0..headerDispatchSheetResponse?.getDispatchSheetEntity()
                                ?.get(i)?.details?.size!! - 1) {
                                Log.e(
                                    "REOS",
                                    "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-headerDispatchSheetResponse?.getDispatchSheetEntity()?.get(i)?.details!!.get(j).invoiceNum" +
                                            headerDispatchSheetResponse?.getDispatchSheetEntity()
                                                ?.get(i)?.details!!.get(j).factura_id
                                )
                                headerDispatchSheetResponse?.getDispatchSheetEntity()
                                    ?.get(i)?.details?.get(j)?.control_id=headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                    .get(i).control_id
                                headerDispatchSheetResponse?.getDispatchSheetEntity()
                                    ?.get(i)?.details?.get(j)?.compania_id=SesionEntity.compania_id
                                headerDispatchSheetResponse?.getDispatchSheetEntity()
                                    ?.get(i)?.details?.get(j)?.fuerzatrabajo_id=SesionEntity.fuerzatrabajo_id
                                headerDispatchSheetResponse?.getDispatchSheetEntity()
                                    ?.get(i)?.details?.get(j)?.usuario_id=SesionEntity.usuario_id

                            }

                        }
                        val executor: ExecutorService = Executors.newFixedThreadPool(1)

                        for (i in 1..1) {
                            executor.execute {
                                println("Tarea $i en ejecución en ${Thread.currentThread().name}")
                                //Thread.sleep(1000)
                                val database by lazy { AppDatabase.getInstance(context.applicationContext) }
                                database?.headerDispatchSheetDao
                                    ?.deleteHeaderDispatchSheet(
                                        FechaDespacho
                                    )
                                database?.headerDispatchSheetDao
                                    ?.insertHeaderDispatchSheet(
                                        headerDispatchSheetResponse?.getDispatchSheetEntity()
                                    )
                                for (i in 0..headerDispatchSheetResponse?.getDispatchSheetEntity()?.size!! - 1) {
                                    database?.detailDispatchSheetDao
                                        ?.deleteDetailDispatchSheet(
                                            headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                                .get(i).control_id
                                        )
                                    database?.detailDispatchSheetDao?.insertDetailDispatchSheet(
                                        headerDispatchSheetResponse?.getDispatchSheetEntity()
                                            ?.get(i)?.details
                                    )
                                }

                                println("Tarea $i completada")
                            }

                        }

                        executor.shutdown()
                        _status.setValue("1")
                        /*Thread {
                            //Do your database´s operations here
                            val database by lazy { AppDatabase.getInstance(context.applicationContext) }
                            database?.headerDispatchSheetDao
                                ?.insertHeaderDispatchSheet(
                                    //headerDispatchSheetResponse?.getDispatchSheetEntity()!!
                                    headerDispatchSheetResponse?.getDispatchSheetEntity()
                                )
                            for (i in 0..headerDispatchSheetResponse?.getDispatchSheetEntity()?.size!! - 1) {
                                database?.detailDispatchSheetDao?.insertDetailDispatchSheet(
                                    headerDispatchSheetResponse?.getDispatchSheetEntity()
                                        ?.get(i)?.details
                                )
                            }

                        }.start()*/


                        Log.e(
                            "REOS",
                            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response.isSuccessful.status" + _status.getValue()
                        )
                    } else {
                        Log.e(
                            "REOS",
                            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response.isSuccessful"
                        )
                        _status.setValue("0")
                        Log.e(
                            "REOS",
                            "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response.isSuccessful.not.status" + _status.getValue()
                        )
                    }
                    Log.e(
                        "REOS",
                        "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-response.noentroenif.status" + _status.getValue()
                    )

                }

                override fun onFailure(call: Call<DispatchSheetResponse?>, t: Throwable) {
                    _status.setValue("0")

                    Log.e(
                        "REOS",
                        "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-error$t"
                    )
                }
            })


        } catch (e: Exception) {
            Log.e(
                "REOS",
                "HeaderDispatchSheetRepositorykt-getStateDispatchSheet-error: " + e.toString()
            )
        }

    }
}

